transform_constantfold.ml

```diff
@@ -26,36 +26,12 @@ let const_fold ({formals; instrs} : analysis_input) : instructions option =
 
   (* Constant fold an expression. *)
   let fold exp =
-    match[@warning "-4"] exp with
-    | Simple _ -> exp
-    | Op (op, [Constant v]) ->
-      begin match op with
-      | Neg -> (Simple (Constant (Int (Eval.value_neg v))))
-      | Not -> (Simple (Constant (Bool (Eval.value_not v))))
-      | Array_index | Array_length -> exp
-      | (Eq | Neq | Lt | Lte | Gt | Gte)
-      | (Plus | Sub | Mult | Div | Mod)
-      | (And | Or) -> assert false
-      end
-    | Op (op, [Constant v1; Constant v2]) ->
-      begin match op with
-      | Eq -> (Simple (Constant (Bool (Eval.value_eq v1 v2))))
-      | Neq -> (Simple (Constant (Bool (Eval.value_neq v1 v2))))
-      | Lt -> (Simple (Constant (Bool (Eval.value_lt v1 v2))))
-      | Lte -> (Simple (Constant (Bool (Eval.value_lte v1 v2))))
-      | Gt -> (Simple (Constant (Bool (Eval.value_gt v1 v2))))
-      | Gte -> (Simple (Constant (Bool (Eval.value_gte v1 v2))))
-      | Plus -> (Simple (Constant (Int (Eval.value_plus v1 v2))))
-      | Sub -> (Simple (Constant (Int (Eval.value_sub v1 v2))))
-      | Mult -> (Simple (Constant (Int (Eval.value_mult v1 v2))))
-      | Div -> (Simple (Constant (Int (Eval.value_div v1 v2))))
-      | Mod -> (Simple (Constant (Int (Eval.value_mod v1 v2))))
-      | And -> (Simple (Constant (Bool (Eval.value_and v1 v2))))
-      | Or -> (Simple (Constant (Bool (Eval.value_or v1 v2))))
-      | (Neg | Not | Array_index | Array_length) -> assert false
-      end
-    | Op (_op, _es) -> exp
-    in
+    let main = {name=""; formals=[]; body=[]} in
+    let prog = {main; functions=[]} in
+    if VarSet.is_empty (expr_vars exp)
+    then Simple (Constant (Eval.eval prog Eval.Heap.empty Eval.Env.empty exp))
```

Comment:
Also I think the code would be more readable with the shorter case first (this is almost always the case): `if not (VarSet.is_empty (expr_vars exp))) ....`.